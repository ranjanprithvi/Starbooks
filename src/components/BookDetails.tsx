import {
    Badge,
    Box,
    Button,
    Divider,
    Flex,
    Grid,
    GridItem,
    Heading,
    Image,
    Skeleton,
    Text,
} from "@chakra-ui/react";
import { Link, useParams } from "react-router-dom";
import useBook from "../hooks/useBook";

const BookDetails = () => {
    const { id } = useParams();

    const { book, isLoading, error } = useBook(id);

    if (error) return <Text>Book Not Found!</Text>;
    return (
        <GridItem colSpan={2} marginBottom={5} maxWidth="1280px" marginX="auto">
            <Box
                marginX={"5"}
                borderColor={"blue.800"}
                borderWidth={"medium"}
                borderRadius={"xl"}
            >
                <Grid
                    width={"100%"}
                    templateRows={{
                        base: "50px 200px repeat(5,50px) 4fr",
                        lg: "50px 200px repeat(5,50px) 4fr",
                    }}
                    templateColumns={{ base: "1fr", md: "1fr 1fr" }}
                    columnGap={{ base: 0, md: 8 }}
                    padding={10}
                >
                    <GridItem rowSpan={7}>
                        {isLoading ? (
                            <Skeleton height="100%"></Skeleton>
                        ) : (
                            <Image
                                src={
                                    book.coverImage ||
                                    "https://bookcart.azurewebsites.net/Upload/Default_image.jpg"
                                }
                                borderRadius={10}
                                // marginX={"auto"}
                                height="100%"
                                // overflow={"clip"}
                            />
                        )}
                    </GridItem>

                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="flex-end"
                            alignItems="center"
                            height="100%"
                        >
                            <Link to={`/books/${id}`}>
                                <Button colorScheme="facebook" marginTop={5}>
                                    Edit
                                </Button>
                            </Link>
                        </Flex>
                    </GridItem>

                    <GridItem>
                        {" "}
                        <Flex alignItems="center" height="100%">
                            <Heading>{book.title}</Heading>
                        </Flex>
                    </GridItem>

                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            alignItems="center"
                            height="100%"
                        >
                            <Text fontWeight="bold">Author: </Text>{" "}
                            <Text>{book.author?.name}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            alignItems="center"
                            height="100%"
                        >
                            <Text fontWeight="bold">Year Published: </Text>{" "}
                            <Text>{book.yearPublished}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            alignItems="center"
                            height="100%"
                        >
                            <Text fontWeight="bold">Genre: </Text>{" "}
                            <Text>{book.genre?.name}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            alignItems="center"
                            height="100%"
                        >
                            <Text fontWeight="bold">Availability: </Text>{" "}
                            {book.numberInStock > 0 ? (
                                <Badge
                                    colorScheme="green"
                                    paddingX={2}
                                    paddingY={{ md: 1 }}
                                >
                                    {`Available (${book.numberInStock} copies)`}
                                </Badge>
                            ) : (
                                <Badge colorScheme="red" padding={2}>
                                    Out of Stock
                                </Badge>
                            )}
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            alignItems="center"
                            height="100%"
                        >
                            <Text fontWeight="bold">Rating: </Text>{" "}
                            <Text>{book.rating}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem
                        marginTop={5}
                        colSpan={{ base: 1, md: 2 }}
                        border="blue.100"
                    >
                        <Heading size={"lg"}>Description</Heading>
                        <Divider marginY="2" />

                        <Text>{book.description}</Text>
                    </GridItem>
                </Grid>
            </Box>
        </GridItem>
    );
};

export default BookDetails;
