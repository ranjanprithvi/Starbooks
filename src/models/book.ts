import { Author } from "./author";
import { Genre } from "./genre";

export interface Book {
    _id: string;
    title: string;
    coverImage: string;
    author: Author;
    genre: Genre;
    rating: number;
    yearPublished: number;
    description: string;
    numberInStock: number;
}

export interface BookQuery {
    author?: string;
    genre?: string;
    search?: string;
    sortBy?: string;
}

export interface BookSort {
    [key: string]: string;
}

export const bookSortFields: BookSort[] = [
    { name: "Date Added", value: "-_id" },
    { name: "Title", value: "title" },
    { name: "Rating", value: "-rating" },
    { name: "Year Published", value: "-yearPublished" },
];
