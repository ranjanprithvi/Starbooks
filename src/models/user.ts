export interface User {
    _id: string;
    name: string;
    email: string;
    membershipExpiry?: Date;
    isAdmin: boolean;
}

export interface UserQuery {
    search?: string;
    sortBy?: string;
    populate?: string;
}
