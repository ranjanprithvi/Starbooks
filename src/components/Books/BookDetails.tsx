import {
    Badge,
    Box,
    Button,
    Divider,
    Flex,
    Grid,
    GridItem,
    Heading,
    Image,
    Skeleton,
    Text,
} from "@chakra-ui/react";
import { Link, useParams } from "react-router-dom";
import useBook from "../../hooks/useBook";
import { BsPencilSquare } from "react-icons/bs";
import { useContext } from "react";
import { LoginContext } from "../../contexts/loginContext";
import { defaultBookCover } from "../../models/book";

const BookDetails = () => {
    const { id } = useParams();
    if (!id) return <Text>Book Not Found!</Text>;
    const { isLoggedIn, isAdmin } = useContext(LoginContext);

    const {
        book: {
            coverImage,
            title,
            numberInStock,
            author,
            genre,
            yearPublished,
            description,
            rating,
            _id,
        },
        isLoading,
        error,
    } = useBook(id);

    if (error) return <Text>Book Not Found!</Text>;
    return (
        <GridItem colSpan={2} marginBottom={5} maxWidth="1280px" marginX="auto">
            <Box
                marginX={"5"}
                borderColor={"blue.800"}
                borderWidth={"medium"}
                borderRadius={"xl"}
            >
                <Grid
                    width={"100%"}
                    templateRows={{
                        base: "50px 4fr repeat(5,50px) auto",
                        lg: "50px 4fr repeat(5,50px) auto",
                    }}
                    templateColumns={{ base: "auto", md: "1fr 1fr" }}
                    columnGap={8}
                    rowGap={{ base: 2, md: 0 }}
                    padding={10}
                >
                    <GridItem rowSpan={7}>
                        {isLoading ? (
                            <Skeleton height="100%"></Skeleton>
                        ) : (
                            <Image
                                src={coverImage || defaultBookCover}
                                borderRadius={10}
                                width={"100%"}
                                // marginX={"auto"}
                                // height="100%"
                                // overflow={"clip"}
                            />
                        )}
                    </GridItem>

                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent={{
                                base: "flex-start",
                                md: "flex-end",
                            }}
                            alignItems="center"
                            height="100%"
                        >
                            {" "}
                            {isLoggedIn && isAdmin && (
                                <Link to={`/books/${id}`}>
                                    <Button
                                        leftIcon={<BsPencilSquare />}
                                        colorScheme="facebook"
                                        marginTop={5}
                                    >
                                        Edit
                                    </Button>
                                </Link>
                            )}
                        </Flex>
                    </GridItem>

                    <GridItem>
                        {" "}
                        <Flex alignItems="center" height="100%">
                            <Heading>{title}</Heading>
                        </Flex>
                    </GridItem>

                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            height="100%"
                            alignItems={{ base: "start", md: "center" }}
                            flexDirection={{ base: "column", md: "row" }}
                        >
                            <Text fontWeight="bold">Author: </Text>{" "}
                            <Text>{author?.name}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            height="100%"
                            alignItems={{ base: "start", md: "center" }}
                            flexDirection={{ base: "column", md: "row" }}
                        >
                            <Text fontWeight="bold">Year Published: </Text>{" "}
                            <Text>{yearPublished}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            height="100%"
                            alignItems={{ base: "start", md: "center" }}
                            flexDirection={{ base: "column", md: "row" }}
                        >
                            <Text fontWeight="bold">Genre: </Text>{" "}
                            <Text>{genre?.name}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            height="100%"
                            alignItems={{ base: "start", md: "center" }}
                            flexDirection={{ base: "column", md: "row" }}
                        >
                            <Text fontWeight="bold">Availability: </Text>{" "}
                            {numberInStock > 0 ? (
                                <Badge
                                    colorScheme="green"
                                    paddingX={2}
                                    paddingY={{ md: 1 }}
                                >
                                    {`Available (${numberInStock} copies)`}
                                </Badge>
                            ) : (
                                <Badge colorScheme="red" padding={2}>
                                    Out of Stock
                                </Badge>
                            )}
                        </Flex>
                    </GridItem>
                    <GridItem>
                        <Divider />
                        <Flex
                            justifyContent="space-between"
                            height="100%"
                            alignItems={{ base: "start", md: "center" }}
                            flexDirection={{ base: "column", md: "row" }}
                        >
                            <Text fontWeight="bold">Rating: </Text>{" "}
                            <Text>{rating}</Text>
                        </Flex>
                    </GridItem>
                    <GridItem
                        marginTop={5}
                        colSpan={{ base: 1, md: 2 }}
                        border="blue.100"
                    >
                        <Heading size={"lg"}>Description</Heading>
                        <Divider marginY="2" />

                        <Text>{description}</Text>
                    </GridItem>
                </Grid>
            </Box>
        </GridItem>
    );
};

export default BookDetails;
