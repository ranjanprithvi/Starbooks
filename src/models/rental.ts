import { Book } from "./book";
import { User } from "./user";

export interface Rental {
    _id: string;
    book: Book;
    user: User;
    dateReturned?: Date;
}
