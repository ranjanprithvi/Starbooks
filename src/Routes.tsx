import React from "react";
import { Navigate, Route, Routes as ReactRoutes } from "react-router-dom";
import Books from "./components/Books/Books";
import BookDetails from "./components/Books/BookDetails";
import BookForm from "./components/Books/BookForm";
import NotFound from "./components/common/NotFound";
import Rentals from "./components/Rentals/Rentals";
import Users from "./components/Users/Users";
import RentalForm from "./components/Rentals/RentalForm";
import UserDetails from "./components/Users/UserDetails";
import UserForm from "./components/Users/UserForm";
import LoginForm from "./components/Users/LoginForm";
import ProtectedAdminComponent from "./components/common/ProtectedAdminComponent";
import ProtectedComponent from "./components/common/ProtectedComponent";
import AuthorGenreList from "./components/Authors/AuthorGenreList";
import Home from "./components/Home";

const Routes = ({ dataView }: { dataView?: string | undefined }) => {
    return (
        <ReactRoutes>
            <Route path="/" element={<Home />} />
            <Route path="/login" element={<LoginForm />} />
            <Route path="/books" element={<Books />} />
            <Route path="bookDetails/:id" element={<BookDetails />} />
            <Route
                path="books/:id"
                element={
                    <ProtectedAdminComponent>
                        <BookForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="users"
                element={
                    <ProtectedAdminComponent>
                        <Users dataView={dataView} />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="userDetails/:id"
                element={
                    <ProtectedComponent>
                        <UserDetails />
                    </ProtectedComponent>
                }
            />
            <Route
                path="users/:id"
                element={
                    <ProtectedAdminComponent>
                        <UserForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals"
                element={
                    <ProtectedAdminComponent>
                        <Rentals />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals/:id"
                element={
                    <ProtectedAdminComponent>
                        <RentalForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route path="/authorGenreList" element={<AuthorGenreList />} />
            <Route path="/not-found" element={<NotFound />} />
            <Route path="*" element={<Navigate to="/not-found" replace />} />
        </ReactRoutes>
    );
};

export default Routes;
