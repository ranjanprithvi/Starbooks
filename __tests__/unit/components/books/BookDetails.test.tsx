import React from "react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";

import BookDetails from "../../../../src/components/Books/BookDetails";
import { defaultBookCover } from "../../../../src/models/book";
import Routes from "../../../../src/Routes";
import { MemoryRouter } from "react-router-dom";

//mock useBook
jest.mock("../../../../src/hooks/useBook");

describe("BookDetails ", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(
                <MemoryRouter initialEntries={[`/bookDetails/123`]}>
                    <Routes />
                </MemoryRouter>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders correctly with no cover image", () => {
        render(
            <MemoryRouter initialEntries={[`/bookDetails/noCover`]}>
                <Routes />
            </MemoryRouter>
        );
        expect(screen.getByRole("img").getAttribute("src")).toBe(
            defaultBookCover
        );
    });

    it("renders error message when no book with given Id is found", () => {
        const errorMsg = "Book Not Found!";

        render(
            <MemoryRouter initialEntries={[`/bookDetails/invalid`]}>
                <Routes />
            </MemoryRouter>
        );

        expect(screen.getByText(errorMsg)).toBeInTheDocument();
    });
});
