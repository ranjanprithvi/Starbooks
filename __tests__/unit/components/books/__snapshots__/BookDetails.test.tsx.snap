// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetails  renders correctly 1`] = `
<p
  className="chakra-text css-0"
>
  Book Not Found!
</p>
`;
